### ISSUE TITLE
'enter' keystroke in code/pre starts new paragraph. expected (for me) behaviour is to extend with "\n" new line

### ISSUE THREAD

**Issue by Suraj151** [ckeditor/ckeditor5-basic-styles#73]
when we enable code and write in it it works fine i.e. it format all written line within code tag but when i hit enter to start with new line then new paragraph starts and code tag ends before it.

can we extend written line (in code/pre tag) with \n (new line) character while enter key and keep it within code/pre tag instead of starts new paragraph.

i mean can we logically detect in which tag cursor is currently operating while hit enter key. if we find it then we can pass "\n" character to input command.

**Comment by jodator** [comment]
We're still missing a [code block plugin](https://github.com/ckeditor/ckeditor5/issues/436). The `Code` plugin from Basic Styles inserts an inline element. It will work the same way as other basic styles in this package - `Bold`, `Italics`, `Underline`, `Strikethrough`. So the <kbd>enter</kbd> key behaviour should be consistent for them.

I'm guessing that you'd like to write longer parts of code like snippets? If so please follow [code blocks feature request](https://github.com/ckeditor/ckeditor5/issues/436).

**Comment by Suraj151** [comment]
Right..!!

my requirement is to support longer code write in editor. and to my bad-luck i don't have enough time as my company allotted only limited time to me for such editor configs.

somehow i able to make pre-tag and media-embed feature for custom ckeditor5-classic-build but now this feature is become a headache (as it need lot work) for me (coz of time limit).

code block feature still under working (i.e. not possible to wait).

@jodator  will you provide me some hint to get me this within short period..! (kind request, as i am in trouble of editor selecting/integrating situation).

**Comment by jodator** [comment]
> somehow i able to make pre-tag and media-embed feature for custom ckeditor5-classic-build but now this feature is become a headache (as it need lot work) for me (coz of time limit).

Could you rephrase this? I'm not sure what you'd like to get - which build, plugins? For media embed - this is already under going so next release should have basic support for this. As for code blocks I think that simple plugin would work - but with very limited behavior.

**Comment by Suraj151** [comment]
i mean i just made my custom build with ckeditor5-build-classic as given in documentation. i have included media embed feature [ckeditor5-embed](https://github.com/Suraj151/ckeditor5-embed) to it. and now i am trying to make **pre tag** code block with bypassing enter within it with \n character.

don't know whether it is possible or not, but i need to keep trying (time limit).

**Comment by jodator** [comment]
OK so **very** limited `<pre>` element support:

```js
class PrePlugin extends Plugin {
	init() {
		const editor = this.editor;
		const schema = editor.model.schema;
		const conversion = editor.conversion;

		schema.register( 'codeBlock', {
			allowWhere: '$block',
			isBlock: true
		} );

		schema.extend( '$text', { allowIn: 'codeBlock' } );

		conversion.elementToElement( {
			model: 'codeBlock',
			view: 'pre'
		} );
	}
}

ClassicEditor
	.create( document.querySelector( '#editor' ), {
		plugins: [ ArticlePluginSet, PrePlugin ]

	} )
	.then( editor => {
		window.editor = editor;
	} )
	.catch( err => {
		console.error( err.stack );
	} );
```

The above code will only:

* convert existing `<pre>` view elements to `codeBlock` model element.
* convert `codeBlock` element back to `<pre>` to the editing view and for data output (`editor.getData()`).
* allow __any__ text content that is allowed in `$block`.
* The <kbd>shift</kbd>+<kbd>enter</kbd> already works - it will insert `<br>` there.

What is missing:

* The `insertCodeBlock` command - (should be fairly simple - just insert `codeBlock` element into the current selection (`writer.insertElement( `codeBlock`, position )`) - rest [similar to the insert table command](https://github.com/ckeditor/ckeditor5-table/blob/cb77e3896d6fccf3cac3b6c1e802f90ecfacdadb/src/commands/inserttablecommand.js#L49-L67).
* The `insertCodeBlock` ui button - should [just call above command](https://github.com/ckeditor/ckeditor5-alignment/blob/11c06c4dc0d0f53a652dd23e06b3ef86b1003f9e/src/alignmentui.js#L136-L157) like `editor.execute( 'insertCodeBlock' );`.
* The <kbd>tab</kbd> key support inside code block (would be nice to intend code - insert spaces/tabs).
* escaping `<`, `>` and `&` in `<pre>` block - right now I don't know what will happen...

Improvements:

* proper styling of code blocks (ie by some attribute that will add class to the `<pre>` element.


![selection_027](https://user-images.githubusercontent.com/247363/43393729-5e2c40f0-93f8-11e8-993b-0198471ec41e.png)

**Comment by jodator** [comment]
ps.: I'm closing this as duplicate of https://github.com/ckeditor/ckeditor5/issues/436. You can ask here or stack overflow (with CKEditor5 tag) if you get stuck on something.

**Comment by Suraj151** [comment]
Thank you..!

i made codeblock plugin as per your suggestion and it working fine now.
